text/0000-stabilize-arbitrary-self-types.md

```diff
@@ -173,107 +170,91 @@ impl MyType {
 
 ## Diagnostics
 
-The existing branches in the compiler for "arbitrary self types" already emit
-excellent diagnostics. We will largely re-use them, with the following improvements:
+The existing branches in the compiler for "arbitrary self types" already emit excellent diagnostics. We will largely re-use them, with the following improvements:
 
-- In the case where a self type is invalid where it doesn't implement `Receiver`,
-  the existing excellent error message will be updated
+- In the case where a self type is invalid where it doesn't implement `Receiver`, the existing excellent error message will be updated.
 - An easy mistake is to implement `Receiver` for `P<T>` without specifying that `T: ?Sized`. `P<Self>` then only works as a `self` parameter in traits `where Self: Sized`, an unusual stipulation. It's not obvious that `Sized`ness is the problem here, so we will identify this case specifically and produce an error giving that hint.
 - There are certain types which feel like they "should" implement `Receiver` but do not: `*const T`, `*mut T`, `Weak` or `NotNull`. If these are encountered as a self type, we should produce a specific diagnostic explaining that they do not implement `Receiver` and suggesting that they could be wrapped in a newtype wrapper if method calls are important. This will require `Weak` and `NonNull` be marked as lang items so that the compiler is aware of the special nature of these types. (The authors of this RFC feel that these extra lang-items _are_ worthwhile to produce these improved diagnostics - if the reader disagrees, please let us know.)
+- Under some circumstances, the compiler identifies method candidates but then discovers that the self type doesn't match. This results currently in a simple "mismatched types" error; we can provide a more specific error message here. The only known case is where a method is generic over `Receiver`, and the caller explicitly specifies the wrong type:
+```rust
+#![feature(receiver_trait)]
+
+use std::ops::Receiver;
+
+struct SmartPtr<'a, T: ?Sized>(&'a T);
+
+impl<'a, T: ?Sized> Receiver for SmartPtr<'a, T> {
+    type Target = T;
+}
+
+struct Foo(u32);
+impl Foo {
+    fn a<R: Receiver<Target=Self>>(self: R) { }
+}
+
+fn main() {
+    let foo = Foo(1);
+    let smart_ptr = SmartPtr(&foo);
+    smart_ptr.a::<&Foo>(); // currently results in "mismatched types"; we can probably do better
+}
+```
```

Comment:
````suggestion
    ```rust
    #![feature(receiver_trait)]

    use std::ops::Receiver;

    struct SmartPtr<'a, T: ?Sized>(&'a T);

    impl<'a, T: ?Sized> Receiver for SmartPtr<'a, T> {
        type Target = T;
    }

    struct Foo(u32);
    impl Foo {
        fn a<R: Receiver<Target=Self>>(self: R) { }
    }

    fn main() {
        let foo = Foo(1);
        let smart_ptr = SmartPtr(&foo);
        smart_ptr.a::<&Foo>(); // currently results in "mismatched types"; we can probably do better
    }
    ```
````